### PR TITLE
Reduce Wires overhead

### PR DESCRIPTION
**Pull request by dwierichs**
**Context:**
`Wires` instances are initialized throughout pennylane in many places, in particular for every gate execution.
Unfortunately this costs a significant amount of time (cProfile showed a relative time consumption of up to 25% 
for the simulation of a 14 qubit circuit with close to 600 two-qubit gates using the qulacs plugin. 

**Description of the Change:**
Moved the uniqueness check from `Wires.__init__` to `_process` in order to perform it only in cases where there 
actually is more than one wire and Wires is not called on a Wires instance.
Changed `Operation.__init__` to only call `Wires` on `wires` if it is not a `Wires` instance yet.
See .github/CHANGELOG.md also (sorry, I was not sure how to reference the pull request before having created it)

**Benefits:**
Efficiency improved.

**Possible Drawbacks:**
The change in wires.py does not duplicate code. The change in operation.py adds one case distinction which 
is well-readable as far as I can tell.

**Related GitHub Issues:**

### CODE REVIEW

**Review by josh146 — COMMENTED** [review]
Thanks for this contribution @dwierichs! This is a great efficiency win. Did you end up running cProfile after the change?

I won't leave a full review at the moment, since it looks like some of the tests are passing. Let us know (by leaving a comment) when the PR is ready for final review.

File: .github/CHANGELOG.md
Comment (dwierichs) [reply .github/CHANGELOG.md:74]: Was not sure how explicit I should be, as I did not work too much on large projects so far.

**Comment by dwierichs** [comment]
I mixed up the numbers, sorry! cProfiling the mentioned simulation with 10 circuit repetitions gave
                  `cumtime` in `Wires.__init__` | Overall runtime   | Wires-time/total time
before PR               2.64s                         |         8.23s          |           31.7 %
after PR                  1.53s                         |         6.97s          |           22.0 %

**Comment by josh146** [comment]
> So I am pretty sure there is more places to save time with calling Wires less but the proposed fix should be the easiest place.

Yep for sure! Looks like we can still do more even after this change to make Wire more efficient :)

**Comment by dwierichs** [comment]
Okay, this is done now. I had not covered the case in `_process` where wires is an `Iterable` but neither are all entries `Wires` nor `Number`s or scalar arrays. This is covered now and the PR should be ready to merge.

File: pennylane/wires.py
Comment (josh146) [inline pennylane/wires.py:41]: How come this is a while loop? From my reading of the below, it seems like, at most, only a single loop will ever execute:

* Either an if statement is activated and the `merged` list is returned, or

* None of the if statements are activated, and the while loop is exited via the `break`.

Unless I'm missing additional logic, could we replace this with an if statement?
```suggestion
    if isinstance(wires, Iterable):
```

File: pennylane/wires.py
Comment (josh146) [inline pennylane/wires.py:55]: Aside from the restructuring of the if statements above, these two lines are the major change, right?

**Review by josh146 — COMMENTED** [review]
Thanks @dwierichs! It looks good. I just left a comment regarding the `while` loop that would be good to get clarified before I approve.

File: pennylane/wires.py
Comment (dwierichs) [reply pennylane/wires.py:55]: Yes, this check was moved to `_process`.

File: pennylane/wires.py
Comment (dwierichs) [reply pennylane/wires.py:41]: One can replace the `while` loop by an `if` loop at the cost of duplicating some code. 
Firstly, one could replace the `break` by a copy of the `WireError` raised at the end of `_process` if no condition applied.
Secondly one could only ever enter the `if isinstance(wires, Iterable)` loop if one of the secondary conditions also applies, resulting in duplication of the check for these secondary conditions (whether all elements are a Wires instance / a scalar).
Thirdly, one could go back to the previous structure of the `if` clauses and duplicate the check for `isinstance(wires, Iterable)`. 
Effectively, the `while` loop is used as an `if` loop in which one can use `break`, which is kind of hacky, I must admit.

File: .github/CHANGELOG.md
Comment (dwierichs) [reply .github/CHANGELOG.md]: This is also taken care of now.

File: pennylane/wires.py
Comment (josh146) [reply pennylane/wires.py:41]: Got it! Thanks @dwierichs.

It's quite a neat solution to avoid the duplicating code. In this case, however, I'm tempted to say that perhaps code readability is more important than avoiding duplicating the exception? On my first reading of the code, I assumed that there was a loop occurring, and when I realized there wasn't, I thought the `while` was a mistake!

Perhaps another approach could be to factor out the `iterable` validation to a separate function:

```python
def _process_iterable(wires):
    if all(isinstance(w, Wires) for w in wires):
        # if the elements are themselves Wires objects, merge them to a new one
        merged = tuple(w for wires_ in wires for w in wires_.tolist())
    elif all(isinstance(w, (str, Number)) or (getattr(w, "shape", None) == tuple()) for w in wires):
        # if the elements are strings or numbers, turn iterable into tuple
        merged = tuple([w.item() if isinstance(w, np.ndarray) else w for w in wires])
    else:
        return None

    if len(set(merged)) != len(merged):
        raise WireError("Wires must be unique; got {}.".format(merged))

    return merged
```

and then, within `_process`,

```python
if isinstance(wires, Iterable):
    wires = _process_iterable(wires)
    if wires is not None:
        return wires

raise WireError(...)
```

Comment (dwierichs) [reply pennylane/wires.py:41]: Okay, I agree it is not as clear as the other solutions. I actually like the "`else return None` --> `if not None return merged`" approach. But maybe we can leave it in `_process`, like so:
```
    if all(isinstance(w, Wires) for w in wires):
        merged = tuple(w for wires_ in wires for w in wires_.tolist())
    elif all(isinstance(w, (str, Number)) or (getattr(w, "shape", None) == tuple()) for w in wires):
        merged = tuple([w.item() if isinstance(w, np.ndarray) else w for w in wires])
    else:
        merged = None
    if merged is not None:
        if len(set(merged)) != len(merged):
            raise WireError("Wires must be unique; got {}.".format(merged))
        return merged
raise WireError(...)
```

Comment (mariaschuld) [reply pennylane/wires.py:41]: Just to understand, is there a speed difference in this rewrite too? 

If so, I like the last version you posted most @dwierichs.

File: pennylane/wires.py
Comment (mariaschuld) [reply pennylane/wires.py:55]: Very clever, how such a small change makes such a difference. Thanks!

**Review by mariaschuld — COMMENTED** [review]
That amazed me @dwierichs, thanks for the PR. It goes to show that we should be more speed conscious when using class data structures! I wonder if there are more improvements of this sort.

Happy to approve, but would also prefer an if statement over the more logically contrived while loop!

**Comment by mariaschuld** [comment]
> I mixed up the numbers, sorry! cProfiling the mentioned simulation with 10 circuit repetitions gave

Wow, 20% in the wires class is still too much. I wonder how we can improve this? The Wires class was supposed to make the core safer, since it guarantees that the different parts only use valid wire labels (i.e. ordered sequences of unique objects). But the overhead seems significant indeed!

File: pennylane/wires.py
Comment (josh146) [reply pennylane/wires.py:41]: I agree, the last version you have @dwierichs is the clearest.

It is also probably faster than the function version I wrote above, since function calls in Python always add a non-negligible overhead.

Comment (dwierichs) [reply pennylane/wires.py:41]: > It is also probably faster than the function version I wrote above, since function calls in Python always add a non-negligible overhead.

Yes, that should be true. 

> Just to understand, is there a speed difference in this rewrite too?

The ordered decisions I made were: 
- remove check for uniqueness from __init__
- therefore have to add it to the cases werer `isinstance(wires, Iterable)` (of which there were two previously)
- therefore unify the two iterable scenarios and write down the test for uniqueness only once.
So this change is part of the speed-up, yes. I hope this answers the question :)

**Comment by dwierichs** [comment]
I agree, it is quite a large overhead. I stumbled upon this because I compared qulacs and pennylane-qulacs and there are orders of magnitude between them for small circuits and still one for the above 14 qubit circuits.
I am lacking the reasoning behind introducing the `Wires` class altogether and to be honest, from the outside it does not look super intuitive to me. 
I think one major improvement would be to cut the wires business our of any execution calls and instead have the user use a `Wires` instance when constructing the circuit.
The entire overhead scales with execution repetitions, which is bad in particular for variational quantum circuits and QML, in my opinion.
Would you be open for a major change in how/where `Wires` is used?

File: pennylane/wires.py
Comment (dwierichs) [reply pennylane/wires.py]: > If so, I like the last version you posted most 

I updated `_process` to this version now.

**Review by josh146 — APPROVED** [review]
Thanks @dwierichs!

**Review by mariaschuld — APPROVED** [review]

**Comment by mariaschuld** [comment]
@dwierichs, of course, we should revisit everything that can improve speed! However, let me add a few thoughts:

* We decided to not make the Wires class user facing, because it is not intuitive enough. I still find it nice to allow users to just have `wires = ['a', 'q1', 2, 3]`, and not worry about the type or some weird class as long as a few minimal properties are fulfilled.
* The abstraction is actually not incidental - as I mentioned, wires in quantum computing are essentially ordered sequences of unique labels, which is really everything we need to address qubits uniquely.
* The wires class was introduced as a standarised bookeeping system that translates between the wire labels a user wants to use, and the wire labels needed to address a device. 
* Before the `Wires` class PennyLane checked for the right logic in wires (i.e. that a 2qubit gate is not applied to the same wire, or that two subset of wires are disjunct). We wanted to collect this logic in the class. We probably don't get around a little overhead from sanity checking.
* We cannot just let users define wire labels directly: For example we once had a bug where someone used a certain kind of integer in a gate that was passed through PennyLane to the device, and the qiskit plugin could not recognise the type. And then of course, we may have devices like hardware chips where a user does not know how to address the wires correctly, for example when the chip has a grid topology addressed by non-consecutive integers. So some kind of translation has to happen.

That's just for context.

Would you mind opening an issue describing the profiling results and possible ways to fix it (if you have ideas)? Thanks so much :)

**Comment by dwierichs** [comment]
Thank you for the detailed overview! I will open an issue once I thought about possible fixes :)
I think there must be a way inbetween, giving the user the same freedom but at the same time not creating a Wires instance after a first circuit compilation step. Thanks for the quick response to this PR!

**Pull request merged**
